dev/generate_matrix.py

```diff
@@ -0,0 +1,209 @@
+
+    Example
+    -------
+    >>> is_valid_version("1.1.0")
+    True
+    >>> is_valid_version("1.1.0rc")
+    False
+    """
+    return re.search(r"^[\.\d]+$", ver) is not None
+
+
+def is_between(ver, min_ver, max_ver):
+    """
+    Returns True if the given version is between the boundary [min_ver, max_ver].
+
+    Example
+    -------
+    >>> is_between("1.1.0", "1.0.0", "1.2.0")
+    True
+    >>> is_between("9.0.0", "1.0.0", "1.2.0")
+    False
+    """
+    v = LooseVersion(ver)
+    return v >= LooseVersion(min_ver) and v <= LooseVersion(max_ver)
+
+
+def _get_latest_micro_versions(versions):
+    """
+    Returns the latest micro versions.
+
+    Example
+    -------
+    >>> versions = [
+    ...     "1.4.2",  # latest in 1.4
+    ...     "1.4.1",
+    ...     "1.4.0",
+    ...     "1.3.1",  # latest in 1.3
+    ...     "1.3.0",
+    ... ]
+    >>> get_latest_micro_versions(versions)
+    ['1.4.2', '1.3.1']
+    """
+    seen = set()
+    res = []
+
+    for ver in sorted(versions, key=LooseVersion, reverse=True):
+        ver = tuple(LooseVersion(ver).version)
+        minor_ver = ver[:2]
+
+        if minor_ver not in seen:
+            seen.add(minor_ver)
+            res.insert(0, ".".join(str(x) for x in ver))
+
+    return res
+
+
+def get_latest_micro_versions(package_name, min_ver, max_ver, excludes=None):
+    """
+    Get the latest micro versions of the specified python package.
+
+    Example
+    -------
+    >>> get_latest_micro_versions("scikit-learn", "0.19.2", "0.23.2")
+    ['0.19.2', '0.20.4', '0.21.3', '0.22.2', '0.23.2']
+    """
+    versions = get_released_versionos(package_name)
+
+    # prevent passing non-existent versions
+    assert min_ver in versions
+    assert max_ver in versions
+
+    if excludes:
+        assert all(v in versions for v in excludes)
+        versions = filter(lambda v: v not in excludes, versions)
+
+    versions = filter(is_valid_version, versions)
+    versions = filter(lambda v: is_between(v, min_ver, max_ver), versions)
+    return list(_get_latest_micro_versions(versions))
+
+
+def get_changed_flavors(diff_files, flavors):
+    """
+    Detects changed flavors from a list of changed files and returns them as a set.
+
+    Example
+    -------
+    >>> get_changed_flavors(["mlflow/xgboost.py"], ["pytorch", "xgboost"])
+    {'xgboost'}
+    """
+    maybe_flavors = set(os.path.splitext(f.split(os.sep)[1])[0] for f in diff_files)
+    return maybe_flavors.intersection(flavors)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--diff-only",
+        action="store_true",
+        help="If specified, ignore unchanged flavors or configurations",
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    dummy_git_diff = """
+mlflow/pytorch/__init__.py
+mlflow/xgboost.py
+"""
+
+    diff_files = dummy_git_diff.strip().split("\n")
+
+    config_pr = read_yaml("ml-package-versions-pr.yml")
```

Comment:
I was under the impression that the `pull_request` cases will check out the associated PR branch and fetch any updates to the `ml-package-versions` yaml that are made by the PR. Perhaps that's not how things work?

Reply:
I was thinking the followin approach.

1. Read `ml-package-versions.yml` on the PR branch
2. Read `ml-package-versions.yml`on the merge target branch (e.g. master)
3. Compare them to detect the difference.
4. Generate a matrix from the diffrence.

Reply:
I'd imagine that we should still run the matrix against all supported minor versions on PRs though, right? If the `ml-package-versions.yml` in the PR branch matches the one `master`, does the lack of a diff imply that we're not going to run any cross-version tests?